Ck/12725 example widget that uses external data 2nd approach (#12788)
ckeditor/ckeditor5: docs/framework/guides/tutorials/data-from-external-source.md

---
category: framework-tutorials
order: 10
---

# Data from external source

> In this tutorial, you will learn how to implement a widget that fetch data from an external source and updates all own instances in set interval of time.

Reviewer: ```suggestion
In this tutorial, you will learn how to implement a widget that fetches data from an external source and updates all own instances in a set interval of time.
```

> You will build an "external data fetch" feature that allows the users to insert predefined widget that will show current Bitcoin rate and it will be updated with the set time interval fetching data from predefined external source. You will use widget utilities and conversion in order to define the behavior of this feature. Later on, you will use button utilities to create a button that will allow for inserting new instances of external source widgets. You will also learn how to update the widget data based on the editor API.

Reviewer: ```suggestion
You will build an "external data fetch" feature that allows users to insert a predefined widget that will show the current Bitcoin rate and it will be updated with the set time interval fetching data from a predefined external source. You will use widget utilities and conversion in order to define the behavior of this feature. Later on, you will use button utilities to create a button that will allow for inserting new instances of external source widgets. You will also learn how to update the widget data based on the editor API.
```

Reviewer: My gut tells me this paragraphs repeats the first paragraph of the guide. Maybe we can get rid of the first paragraph then?

Reviewer: > use button utilities

I'd call it a UI library. And link to the guide [https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/framework/guides/architecture/ui-library.html.](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/framework/guides/architecture/ui-library.html.)

I couldn't find a reference to the `ButtonView` in the API in the whole guide either.

Author: > I couldn't find a reference to the `ButtonView` in the API in the whole guide either.

Correct, now i've update it.

								themeImporter: {
									themePath: require.resolve( '@ckeditor/ckeditor5-theme-lark' )
								},
								minify: true
							} )
						}
					}
				]
			}
		]
	},

	// Useful for debugging.
	devtool: 'source-map',

	// By default webpack logs warnings if the bundle is bigger than 200kb.
	performance: { hints: false }
};
```

Add an `index.html` page:

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<title>CKEditor 5 Framework – Implementing a simple widget</title>
	</head>
	<body>
		<div id="editor">
			<p>Editor content goes here.</p>
		</div>

		<script src="dist/bundle.js"></script>
	</body>
</html>
```

The application entry point (`app.js`):

```js
// app.js

import ClassicEditor from '@ckeditor/ckeditor5-editor-classic/src/classiceditor';
import Essentials from '@ckeditor/ckeditor5-essentials/src/essentials';
import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
import Heading from '@ckeditor/ckeditor5-heading/src/heading';
import List from '@ckeditor/ckeditor5-list/src/list';
import Bold from '@ckeditor/ckeditor5-basic-styles/src/bold';
import Italic from '@ckeditor/ckeditor5-basic-styles/src/italic';

> import ExternalWidget from './external-widget/externalwidget';

Reviewer: When I first saw that class name I was confused and I thought this is part of our API. What about `ExternalDataWidget` to be more precise?

Author: Agree :slightly\_smiling\_face: , done: 3061b45

import ExternalWidgetEditing from './externalwidgetediting';
import ExternalWidgetUI from './externalwidgetui';

export default class ExternalWidget extends Plugin {
	static get requires() {
		return [ ExternalWidgetEditing, ExternalWidgetUI ];
	}
}
```

The UI part (empty for now):

```js
// external-widget/externalwidgetui.js

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';

export default class ExternalWidgetUI extends Plugin {
	init() {
		console.log( 'ExternalWidgetUI#init() got called' );
	}
}
```

And the editing part (empty for now):

```js
// external-widget/externalwidgetediting.js

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';

export default class ExternalWidgetEditing extends Plugin {
	init() {
		console.log( 'ExternalWidgetEditing#init() got called' );
	}
}
```

At this stage you can build the project and open it in the browser to verify if it is building correctly.

Use this command `./node_modules/.bin/webpack --mode development` in root folder of the widget to build the project.

After the build is completed, open `index.html` in your browser to check if all is correct at this stage.

## The model and the view layers

> The external widget feature will be {@link module:engine/model/schema~SchemaItemDefinition defined as an inline} (text-like) element so it will be inserted into other editor blocks, like `<paragraph>`, that allow text. The external widget will have a `data-resource-url` attribute. This means that the model containing some text and a external widget will look like this:

Reviewer: ```suggestion
The external widget feature will be {@link module:engine/model/schema~SchemaItemDefinition defined as an inline} (text-like) element so it will be inserted into other editor blocks that allow text e.g. `<paragraph>`. The external widget will have a `data-resource-url` attribute. This means that the model containing some text and an external widget will look like this:
```

```html
<paragraph>

> 	External value: <externalElement data-resource-url="RESOURCE_URL"></externalElement>.

Reviewer: Using this syntax that is neither our model nor HTML will create a lot of confusion unless you explain what it is. This is the syntax used by our test tools to make our lives a bit easier but people who read this will probably assume that this is the HTML you get in editor output (`editor.getData()`). You can explain this by mentioning the [model section of the engine guide](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/framework/guides/architecture/editing-engine.html#model) and/or the [CKEditor inspector](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/framework/guides/development-tools.html#ckeditor-5-inspector).

Author: WDYT? 1d84f40

</paragraph>
```

### Defining the schema

> The schema definition in this widget is almost the same as in {@link framework/guides/tutorials/implementing-an-inline-widget#defining-the-schema inline widget tutorial}, the only thing that is different is in `allowAttributes`, in our case we want to allow `'data-resource-url'` attribute.

Reviewer: ```suggestion
The schema definition of this widget is almost the same as in {@link framework/guides/tutorials/implementing-an-inline-widget#defining-the-schema inline widget tutorial}, the only thing that is different is in `allowAttributes`, in our case we want to allow the `'data-resource-url'` attribute.
```

> Instead of passing all the attributes to the config object we can use a {@link framework/guides/deep-dive/schema#generic-items generic items} to inherit already predefined options.

Reviewer: ```suggestion
Instead of passing all the attributes to the config object we can use {@link framework/guides/deep-dive/schema#generic-items generic items} to inherit already predefined options.
```

> You will also use this opportunity to import the theme file (`theme/externalwidget.css`).

Reviewer: ```suggestion
You can also use this opportunity to import the theme file (`theme/externalwidget.css`).
```

?

```js
// external-widget/externalwidgetediting.js

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';

import './theme/externalwidget.css';                                           // ADDED

export default class ExternalWidgetEditing extends Plugin {
	init() {
		console.log( 'ExternalWidgetEditing#init() got called' );

		this._defineSchema();                                                  // ADDED
	}

	_defineSchema() {                                                          // ADDED
		const schema = this.editor.model.schema;

		schema.register( 'externalElement', {
			// Inheriting all from the generic item
			inheritAllFrom: '$inlineObject',

			// The external widget can have many attributes
			allowAttributes: [ 'data-resource-url' ]
		} );
	}
}
```

The schema is defined so now you can define the model-view converters.

### Defining converters

> The HTML structure (data output) of the converter will be a `<span>` with a `data-resource-url` attribute with external resource url as a value.

Reviewer: ```suggestion
The HTML structure (data output) of the converter will be a `<span>` with a `data-resource-url` attribute with the external resource url as a value.
```

		this._defineConverters();                                              // ADDED
	}

	_defineSchema() {                                                          // ADDED
		// ...
	}

	_defineConverters() {                                                      // ADDED
		const editor = this.editor;

		editor.conversion.for( 'upcast' ).elementToElement( {
			view: {
				name: 'span',
				attributes: [ 'data-resource-url' ]
			},
			model: ( viewElement, { writer } ) => {
				const externalUrl = viewElement.getAttribute( 'data-resource-url' );

				return writer.createElement( 'externalElement', {
					'data-resource-url': externalUrl
				} );
			}
		} );

		editor.conversion.for( 'dataDowncast' ).elementToElement( {
			model: 'externalElement',
			view: ( modelElement, { writer } ) => {
				return writer.createEmptyElement( 'span', {
					'data-resource-url': modelElement.getAttribute( 'data-resource-url' )
				} );
			}
		} );

		editor.conversion.for( 'editingDowncast' ).elementToElement( {
			model: 'externalElement',
			view: ( modelElement, { writer } ) => {
				const externalDataPreviewElement = writer.createRawElement( 'span', null, function( domElement ) {

> 					// for now show some static text

Reviewer: ```suggestion
					// For now show some static text
```

					domElement.textContent = 'Data placeholder';
				} );

				const externalWidgetContainer = writer.createContainerElement( 'span', null, externalDataPreviewElement );

				return toWidget( externalWidgetContainer, writer, {
					label: 'External widget'
				} );
			}
		} );
	}
}
```

### Feature styles

> As you could notice, the editing part imports the `./theme/externalwidget.css` CSS file which describes how the widget will be animated when new value arrives:

Reviewer: ```suggestion
As you might have noticed, the editing part imports the `./theme/externalwidget.css` CSS file which describes how the widget will be animated when new value arrives:
```

```css
/* external-widget/theme/externalwidget.css */

.external-widget-bounce {
	animation: external-widget-bounce-animation 0.7s 1;
}

@keyframes external-widget-bounce-animation {
	0% {
		box-shadow: 0px 0px 0px 0px rgba(116, 60, 205, 1);
	}

	100% {
		box-shadow: 0px 0px 0px 7px rgba(116, 60, 205, 0);
	}
}
```

### Command

> The {@link framework/guides/architecture/core-editor-architecture#commands command} for the external widget feature will insert an `<externalElement>` element (if allowed by the schema) at the selection and after insert is completed the selection will be set on inserted widget.

Reviewer: ```suggestion
The {@link framework/guides/architecture/core-editor-architecture#commands command} for the external widget feature will insert an `<externalElement>` element (if allowed by the schema) at the selection and set the selection on the inserted widget.
```

import ExternalWidgetCommand from './externalwidgetcommand';                   // ADDED

import './theme/externalwidget.css';

export default class ExternalWidgetEditing extends Plugin {
	static get requires() {
		return [ Widget ];
	}

	init() {
		console.log( 'ExternalWidgetEditing#init() got called' );

		this._defineSchema();
		this._defineConverters();

		// ADDED
		this.editor.commands.add( 'external', new ExternalWidgetCommand( this.editor ) );
	}

	_defineSchema() {
		// ...
	}

	_defineConverters() {
		// ...
	}
}
```

## Creating the UI

The UI part will provide a button that user can click to insert external widget into the editor.

```js
// external-widget/externalwidgetui.js

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
import { ButtonView } from '@ckeditor/ckeditor5-ui/src';

class ExternalWidgetUI extends Plugin {
	init() {
		const editor = this.editor;
		const externalWidgetCommand = editor.commands.get( 'external' );

		// The "external" button must be registered among the UI components of the editor
		// to be displayed in the toolbar.
		editor.ui.componentFactory.add( 'external', locale => {
			const button = new ButtonView( locale );

> 			button.set( {

Reviewer: We could use the official icon. It's public domain https://en.bitcoin.it/wiki/Promotional_graphics.

Author: Good idea! :+1:  Added: 1712467

				label: 'Bitcoin rate',
				tooltip: true,
				withText: true
			} );

			// Disable the external widget button when the command is disabled.
			button.bind( 'isEnabled' ).to( externalWidgetCommand );

			// Execute the command when the button is clicked (executed).
			button.on( 'execute', () => {
				editor.execute( 'external' );

> 				// set focus on the editor content

Reviewer: ```suggestion
				// Set focus on the editor content
```

import Heading from '@ckeditor/ckeditor5-heading/src/heading';
import List from '@ckeditor/ckeditor5-list/src/list';
import Bold from '@ckeditor/ckeditor5-basic-styles/src/bold';
import Italic from '@ckeditor/ckeditor5-basic-styles/src/italic';

import ExternalWidgetCommand from './externalwidgetcommand';

ClassicEditor
	.create( document.querySelector( '#editor' ), {
		plugins: [ Essentials, Paragraph, Heading, List, Bold, Italic, ExternalWidget ],

		// Insert the "external" button into the editor toolbar.
		toolbar: [ 'heading', 'bold', 'italic', 'numberedList', 'bulletedList', '|', 'external', '|', 'undo', 'redo' ]
	} )
	.then( editor => {
		console.log( 'Editor was initialized', editor );

		// Expose for playing in the console.
		window.editor = editor;
	} )
	.catch( error => {
		console.error( error.stack );
	} );
```

## Handling the external source

In this tutorial we will use an external API that provides a current Bitcoin rate (in USD). Endpoint that we will use do not need any API keys and it's free to use but it has some [limitations](https://binance-docs.github.io/apidocs/spot/en/#limits).

```js
'https://api2.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT'
```

> Data will be fetched in with time interval set to 15 seconds. Every instance of the widget will be updated at once in same time. To achieve that we need to modify our `ExternalWidgetEditing` class.

Reviewer: ```suggestion
Data will be fetched every 15 seconds. Every instance of the widget will be updated at the same time. To achieve that, we need to modify our `ExternalWidgetEditing` class.
```

```js

class ExternalWidgetEditing extends Plugin {
	//
	constructor( editor ) {

> 		// the default constructor calls the parent constructor

Reviewer: ```suggestion
		// The default constructor calls the parent constructor
```

		super( editor );

> 		// property that keep the interval id

Reviewer: ```suggestion
		// Property that keep the interval id
```

		this.intervalId = this._intervalFetch();

> 		// last fetched value

Reviewer: ```suggestion
		// Last fetched value
```

		this.externalDataValue = '';
	}

	static get requires() {
		return [ Widget ];
	}

> 	// this method will help us to clear the interval

Reviewer: ```suggestion
	// This method will help us to clear the interval
```

	destroy() {
		clearInterval( this.intervalId );
	}

	init() {
		this._defineSchema();
		this._defineConverters();

> 		// initial execute function to fetch and update the data

Reviewer: ```suggestion
		// Initial execute function to fetch and update the data
```

		this._updateWidgetData();

		this.editor.commands.add( 'external', new ExternalWidgetCommand( this.editor ) );
	}

> 	// interval function

Reviewer: ```suggestion
	// Interval function
```

	_intervalFetch() {
		return setInterval( () => this._updateWidgetData(), 15000 ); // set time interval to 15s
	}

> 	// fetch data and update all widget instances

Reviewer: ```suggestion
	// Fetch data and update all widget instances
```

	async _updateWidgetData( externalUrl = RESOURCE_URL ) {
		try {
			const response = await fetch( externalUrl );
			const data = await response.json();
			const updateTime = new Date( data.closeTime );

> 			// example parsed data: $17098.35 - 09/11/2022, 18:04:18

Reviewer: ```suggestion
			// Example parsed data: $17098.35 - 09/11/2022, 18:04:18
```

			const parsedData = '$' + Number( data.lastPrice ).toFixed( 2 ) + ' - ' + updateTime.toLocaleString();

> 			// update property with last fetched and parsed data

Reviewer: ```suggestion
			// Update property with last fetched and parsed data
```

			this.externalDataValue = parsedData;

			const rootElement = this.editor.model.document.getRoot();

> 			// loop over whole editor content, search for external widget instances

Reviewer: ```suggestion
			// Iterate over whole editor content, search for external widget instances
```

			// and trigger `recovertItem` function
			for ( const { item } of this.editor.model.createRangeIn( rootElement ) ) {
				if ( item.is( 'element', 'externalElement' ) ) {
					this.editor.editing.reconvertItem( item );
				}
			}
		} catch ( error ) {
			console.error( error );
		}
	}

	_defineSchema() {
		// ...
	}

	_defineConverters() {
		// ...

		editor.conversion.for( 'editingDowncast' ).elementToElement( {
			model: 'externalElement',
			view: ( modelElement, { writer } ) => {
				const externalValueToShow = this.externalDataValue;

				const externalDataPreviewElement = writer.createRawElement( 'span', null, function( domElement ) {

> 					// when the value is not present (initial run) show a placeholder

Reviewer: ```suggestion
					// When the value is not present (initial run) show a placeholder
```

					domElement.textContent = externalValueToShow || 'Fetching data...';

> 					// if new value arrive add a css animation effect to show that data were updated

Reviewer: ```suggestion
					// If a new value arrives, add a CSS animation effect to show that data were updated
```

					if ( externalValueToShow ) {
						domElement.classList.add( 'external-widget-bounce' );

> 						// remove animation class when it ends

Reviewer: ```suggestion
						// Remove the animation class when it ends
```

						setTimeout( () => domElement.classList.remove( 'external-widget-bounce' ), 1100 );
					}
				} );

				const externalWidgetContainer = writer.createContainerElement( 'span', null, externalDataPreviewElement );

				return toWidget( externalWidgetContainer, writer, {
					label: 'External widget'
				} );
			}
		} );
	}
}

```

> Traversing over the whole editor content can be tricky. When there relatively little content this method is sufficient, but if there is a lot of content, we suggest to use `WeakMap`.

Reviewer: Saying it's tricky is a colloquialism. Traversal can be challenging.

Author: Resolved by @godai78

> Traversing over the whole editor content can be tricky. When there relatively little content this method is sufficient, but if there is a lot of content, we suggest to use `WeakMap`.

Reviewer: ```suggestion
Editor content traversal can be a challenging process. The presented method is sufficient when there is relatively little content. Otherwise, a `WeakMap` will be a better option.
```

					// If a new value arrives, add a CSS animation effect to show that data were updated
					if ( externalValueToShow ) {
						domElement.classList.add( 'external-data-widget-bounce' );
						// Remove the animation class when it ends
						setTimeout( () => domElement.classList.remove( 'external-data-widget-bounce' ), 1100 );
					}
				} );

				const externalWidgetContainer = writer.createContainerElement( 'span', null, externalDataPreviewElement );

				return toWidget( externalWidgetContainer, writer, {
					label: 'External widget'
				} );
			}
		} );
	}
}

```

The editor content traversal can be a challenging process. The presented method is sufficient when there is relatively little content. Otherwise, a `WeakMap` will be a better option.

## Demo

You can see the external data widget implementation in action in the editor below.

{@snippet framework/tutorials/external-data-widget}

## Final solution

The following code snippet contains the complete implementation of the `ExternalDataWidget` plugin (and all of its dependencies) and the code needed to run the editor. You can paste it into the `app.js` file and it will run out–of–the–box (excluded the Bitcoin logo):

> ```js

Author: Removing an icon from this fully worked snippet is done on purpose (small mention about it is at the end: `(excluded the Bitcoin logo)`).  
Like it is written its a `copy/paste` code example. When import of the svg icon will be added - it will cause errors when user wont have that icon in his directory. (also the icon is removed from theButtonView - left `withText: true`)

Of course i can add it, but first can mention about it above the code snippet, so the user needs to be sure that the icon is there where the import is made from.

WDYT??
